test/dotnet/input/InsertBulk/insertBulkErrors.txt

```diff
@@ -0,0 +1,322 @@
+##########################################################
+#################### TEST DETAILS ########################
+### 1. Testing explicit transaction (error case handled in 5.)
+###    a. Commit without error
+###    b. Rollback without error
+### 2. Index with without transaction
+### 3. Primary Key error case
```

Comment:
we should test each error case without transaction, with transaction followed by commit and rollback
we should also add tests for savepoints and rollback to savepoint, without error and with error
we should manually test both error cases where savepoint state is not as expected
if supported, test with implicit transaction on/off, xact abort on/off
I am not clear how we are testing reset connection with pooling
how are we testing out of order packet error?

Reply:
> we should test each error case without transaction, with transaction followed by commit and rollback

Thats already done

> we should also add tests for savepoints and rollback to savepoint, without error and with error
I will add tests for manual savepoints as well

> we should manually test both error cases where savepoint state is not as expected
That cannot be tested from InsertBulk Copy point of view, right?

> if supported, test with implicit transaction on/off, xact abort on/off
Let me check

> I am not clear how we are testing reset connection with pooling
These are the test cases:
https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2887/files#diff-b1b7adfb8ac542c63531c143322672a0ec670b83ace04a84a53eec81e0e562c1R293
Also check PR description for more validation on these tests through logs

> how are we testing out of order packet error?
.NET driver sends attention to interrupt any Bulk Load Error in a rare case where the error has already been deducted but there are rows the driver has still not sent in the request. For example, This we encounter in the NOT NULL case where first row is NULL - the driver gets error and it sends an out of band attention packet for which it does not wait for a response to close the connection

Reply:
There is a .net driver level option to useInternalTransaction: https://learn.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlbulkcopyoptions?view=netframework-4.8.1&redirectedfrom=MSDN

But in that .net driver itself starts an explicit transaction